### ISSUE TITLE
About method DifferentiableAugment

### ISSUE THREAD

**Issue by DanceInDark** [VDIGPKU/DADA#10]
Hi, thanks for the good work. I have one question need your help. I am confused with the method DifferentiableAugment in https://github.com/VDIGPKU/DADA/blob/master/search_relax/model_search.py#L43. 
`class DifferentiableAugment(nn.Module):
    def __init__(self, sub_policy):
        super(DifferentiableAugment, self).__init__()
        self.sub_policy = sub_policy

    def forward(self, origin_images, probability_b, magnitude):
        images = origin_images
        adds = 0
        for i in range(len(self.sub_policy)):
            if probability_b[i].item() != 0.0:
                images = images - magnitude[i]
                adds = adds + magnitude[i]
        images = images.detach() + adds
        return images`
It seems that image processing oprations are not really called to preprocess training images on search stage. For each sub_policy, just minus the magnitude? If I misunderstood something, please tell me, thanks.

**Comment by latstars** [comment]
The image processing operations are called as this code https://github.com/VDIGPKU/DADA/blob/master/search_relax/dataset.py#L68. The code in model_search.py is just to support the gradient calculation.
`    def __getitem__(self, index):
        if self.search:
            # start_time = time.time()
            img, target = self.dataset.__getitem__(index)
            img = self.pre_transforms(img)
            magnitude = self.magnitudes.clamp(0, 1)[self.weights_index.item()]
            sub_policy = self.ops_names[self.weights_index.item()]
            probability_index = self.probabilities_index[self.weights_index.item()]
            image = img
            for i, ops_name in enumerate(sub_policy):
                if probability_index[i].item() != 0.0:
                    image = apply_augment(image, ops_name, magnitude[i])
            image = self.after_transforms(image)
            return image, target`

**Comment by DanceInDark** [comment]
> The image processing operations are called as this code https://github.com/VDIGPKU/DADA/blob/master/search_relax/dataset.py#L68. The code in model_search.py is just to support the gradient calculation.
> 
> ```
> def __getitem__(self, index):
>         if self.search:
>             # start_time = time.time()
>             img, target = self.dataset.__getitem__(index)
>             img = self.pre_transforms(img)
>             magnitude = self.magnitudes.clamp(0, 1)[self.weights_index.item()]
>             sub_policy = self.ops_names[self.weights_index.item()]
>             probability_index = self.probabilities_index[self.weights_index.item()]
>             image = img
>             for i, ops_name in enumerate(sub_policy):
>                 if probability_index[i].item() != 0.0:
>                     image = apply_augment(image, ops_name, magnitude[i])
>             image = self.after_transforms(image)
>             return image, target
> ```

I see. Thank you.

**Comment by DanceInDark** [comment]
> The image processing operations are called as this code https://github.com/VDIGPKU/DADA/blob/master/search_relax/dataset.py#L68. The code in model_search.py is just to support the gradient calculation.
> 
> ```
> def __getitem__(self, index):
>         if self.search:
>             # start_time = time.time()
>             img, target = self.dataset.__getitem__(index)
>             img = self.pre_transforms(img)
>             magnitude = self.magnitudes.clamp(0, 1)[self.weights_index.item()]
>             sub_policy = self.ops_names[self.weights_index.item()]
>             probability_index = self.probabilities_index[self.weights_index.item()]
>             image = img
>             for i, ops_name in enumerate(sub_policy):
>                 if probability_index[i].item() != 0.0:
>                     image = apply_augment(image, ops_name, magnitude[i])
>             image = self.after_transforms(image)
>             return image, target
> ```

emm，能不能加个微信或qq，想请教你一些问题

**Comment by DanceInDark** [comment]
> The image processing operations are called as this code https://github.com/VDIGPKU/DADA/blob/master/search_relax/dataset.py#L68. The code in model_search.py is just to support the gradient calculation.
> 
> ```
> def __getitem__(self, index):
>         if self.search:
>             # start_time = time.time()
>             img, target = self.dataset.__getitem__(index)
>             img = self.pre_transforms(img)
>             magnitude = self.magnitudes.clamp(0, 1)[self.weights_index.item()]
>             sub_policy = self.ops_names[self.weights_index.item()]
>             probability_index = self.probabilities_index[self.weights_index.item()]
>             image = img
>             for i, ops_name in enumerate(sub_policy):
>                 if probability_index[i].item() != 0.0:
>                     image = apply_augment(image, ops_name, magnitude[i])
>             image = self.after_transforms(image)
>             return image, target
> ```


I have another question, Where are the declarations of self.weights_index at https://github.com/VDIGPKU/DADA/blob/master/search_relax/dataset.py#L73 and self.probabilities_index at https://github.com/VDIGPKU/DADA/blob/master/search_relax/dataset.py#L75?

**Comment by latstars** [comment]
> emm，能不能加个微信或qq，想请教你一些问题

好的，我的微信号latstars
